doc/features/speculative-executions/README.md

```diff
@@ -0,0 +1,214 @@
+  |                |     server error    |
+  |                |<--------------------|
+  |                |              |
+  |                |   retry on host3
+  |                |-------------------->|
+  |                |              |      |
+  |                | host1 replies|      |
+  |                |<-------------|      |
+  |   complete     |                     |
+  |<---------------|                     |
+  |                | cancel              |
+  |                |-------------------->|
+```
+
+The only impact is that all executions of the same query always share the same query plan, so each host is used by at
+most one execution.
+
+## Tuning and practical details
+ 
+The goal of speculative executions is to improve overall latency (the time between `execute(query)` and `complete` in
+the diagrams above) at high percentiles. On the flipside, they cause the driver to send more individual requests, so
+throughput does not necessarily improve.
+
+One side-effect of speculative executions is that many requests are cancelled, which can lead to a phenomenon called
+stream id exhaustion: each TCP connection can handle multiple simultaneous requests, identified by a unique
+number called stream id. When a request gets cancelled, we can't reuse its stream id immediately because we might
+still receive a response from the server later. If this happens often, the number of available stream ids diminishes
+over time, and when it goes below a given threshold we close the connection and create a new one. If requests are often
+cancelled, so will see connections being recycled at a high rate.
+
+This problem is more likely to happen with old server versions (Apache Cassandra version 2.0 or below and DSE 4.6 or 
+below) which only support version 1 and 2 of the  native protocol where each TCP connection only has 128 available
+stream ids. With modern server versions, there are 32K stream ids per connection, so higher cancellation rates can be 
+sustained.
+
+Another issue that might arise is that you get unintuitive results because of request ordering. Suppose you run the
+following query with speculative executions enabled:
+
+```
+insert into my_table (k, v) values (1, 1);
+```
+
+The first execution is a bit too slow, so a second execution gets triggered. Finally, the first execution completes,
+so the client code gets back an acknowledgement, and the second execution is cancelled. However, cancelling only means
+that the driver stops waiting for the server's response, the request could still be on the wire; let us assume that
+this is the case. Now you run the following query, which completes successfully:
+
+```
+delete from my_table where k = 1;
+```
+
+But now the second execution of the first query finally reaches its target node, which applies the mutation. The row
+that you've just deleted is back!
+
+**Using [query timestamps](../query-timestamps)**, which are enabled by default, prevents
```

Comment:
That reminds me that it would be good to validate that if using client timestamps, that the same timestamp is used among all executions.  Can't imagine that not being the case, but will double check.